File: src/lib/server/integrations/notion.service.ts
import { logger } from '$lib/server/utils/logger';

/**
 * NotionService
 * - Abstraksi tipis di atas Notion API untuk kebutuhan platform.
 * - Menggunakan dynamic import agar tidak memaksa dependency saat runtime jika belum terpasang.
 *   Pasang paket: `npm i @notionhq/client`
 */
export class NotionService {
  private token: string;

  constructor(token?: string) {
    this.token = token || process.env.NOTION_API_KEY || '';
    if (!this.token) {
      logger.warn('NOTION_API_KEY is not set. NotionService is disabled.');
    }
  }

  private async getClient() {
    if (!this.token) throw new Error('NOTION_API_KEY is not configured');
    // Dynamic import to avoid hard dependency if package is not installed yet
    const mod = await import('@notionhq/client');
    const { Client } = mod as unknown as { Client: new (opts: { auth: string }) => any };
    return new Client({ auth: this.token });
  }

  /**
   * Cari proyek yang sedang membuka rekrutmen pada database Notion.
   * Konvensi default: properti select bernama "Status" bernilai "Open".
   * Jika filter gagal (karena skema berbeda), akan fallback ke hasil tanpa filter (maks 25).
   */
  async findOpenProjects(databaseId: string) {
    const client = await this.getClient();

    try {
      const res = await client.databases.query({
        database_id: databaseId,
        filter: {
          property: 'Status',
          select: { equals: 'Open' }
        },
        page_size: 50
      });
      return res.results.map((page: any) => this.mapProject(page));
    } catch (err) {
      logger.warn('NotionService.findOpenProjects failed with filter. Fallback to unfiltered.', { err });
      const res = await client.databases.query({ database_id: databaseId, page_size: 25 });
      return res.results.map((page: any) => this.mapProject(page));
    }
  }

  /**
   * Ambil konten dokumen dari database Notion berdasarkan judul halaman (Name/Title).
   * Jika properti bernama berbeda, gunakan filter custom sesuai kebutuhan.
   */
  async getDocumentByName(databaseId: string, docName: string): Promise<{ pageId: string; title: string; text: string } | null> {
    const client = await this.getClient();

    // Coba filter properti Title bernama 'Name'
    const query = async () => {
      try {
        return await client.databases.query({
          database_id: databaseId,
          filter: {
            property: 'Name',
            title: { equals: docName }
          },
          page_size: 1
        });
      } catch (err) {
        // Fallback: tanpa filter, lalu cari secara manual
        logger.warn('NotionService.getDocumentByName failed Title filter. Fallback to manual search.', { err });
        const res = await client.databases.query({ database_id: databaseId, page_size: 100 });
        return {
          results: res.results.filter((p: any) => this.extractTitle(p).trim().toLowerCase() === docName.trim().toLowerCase())
        } as any;
      }
    };

    const res = await query();
    const page = (res.results && res.results[0]) || null;
    if (!page) return null;

    const pageId = page.id;
    const title = this.extractTitle(page);
    const text = await this.getPagePlainText(client, pageId);
    return { pageId, title, text };
  }

  // Helpers
  private mapProject(page: any) {
    const title = this.extractTitle(page);
    const tags = this.extractMultiSelect(page, 'Tags') || this.extractMultiSelect(page, 'Kategori') || [];
    const status = this.extractSelect(page, 'Status');
    return {
      id: page.id,
      title,
      status,
      tags
    };
  }

  private extractTitle(page: any): string {
    const props = page.properties || {};
    const titleProp = props.Name || props.Title || props.Judul;
    if (!titleProp || !Array.isArray(titleProp.title)) return '';
    return titleProp.title.map((t: any) => t.plain_text || '').join('').trim();
  }

  private extractSelect(page: any, propName: string): string | null {
    const props = page.properties || {};
    const prop = props[propName];
    if (!prop || !prop.select) return null;
    return prop.select?.name || null;
  }

  private extractMultiSelect(page: any, propName: string): string[] | null {
    const props = page.properties || {};
    const prop = props[propName];
    if (!prop || !Array.isArray(prop.multi_select)) return null;
    return prop.multi_select.map((s: any) => s.name).filter(Boolean);
  }

  private async getPagePlainText(client: any, pageId: string): Promise<string> {
    const blocks: string[] = [];
    const appendBlocks = async (pid: string) => {
      let cursor: string | undefined = undefined;
      do {
        const res: any = await client.blocks.children.list({ block_id: pid, start_cursor: cursor });
        for (const b of res.results as any[]) {
          blocks.push(this.blockToText(b));
          if (b.has_children) {
            await appendBlocks(b.id);
          }
        }
        cursor = res.has_more ? res.next_cursor : undefined;
      } while (cursor);
    };

    await appendBlocks(pageId);
    return blocks.filter(Boolean).join('\n');
  }

  private blockToText(block: any): string {
    const rt = (rts: any[]) => (rts || []).map((r) => r.plain_text || '').join('');
    const t = block.type;
    const data = block[t] || {};
    switch (t) {
      case 'heading_1':
        return `# ${rt(data.rich_text)}`;
      case 'heading_2':
        return `## ${rt(data.rich_text)}`;
      case 'heading_3':
        return `### ${rt(data.rich_text)}`;
      case 'paragraph':
        return rt(data.rich_text);
      case 'bulleted_list_item':
        return `- ${rt(data.rich_text)}`;
      case 'numbered_list_item':
        return `1. ${rt(data.rich_text)}`;
      case 'to_do':
        return `- [${data.checked ? 'x' : ' '}] ${rt(data.rich_text)}`;
      case 'quote':
        return `> ${rt(data.rich_text)}`;
      case 'callout':
        return rt(data.rich_text);
      case 'code':
        return '```\n' + (data?.rich_text?.[0]?.plain_text || '') + '\n```';
      case 'toggle':
        return rt(data.rich_text);
      default:
        return '';
    }
  }
}

export const notionService = new NotionService();

File: backend/notion_service.py
from __future__ import annotations
from typing import List, Optional, Dict, Any
from notion_client import Client
from .config import settings

class NotionService:
    def __init__(self) -> None:
        if not settings.NOTION_TOKEN:
            raise RuntimeError("NOTION_TOKEN not set.")
        self.client = Client(auth=settings.NOTION_TOKEN)
        self.db_projects = settings.NOTION_DB_PROJECTS_ID
        self.db_docs = settings.NOTION_DB_DOCS_ID
        if not self.db_projects or not self.db_docs:
            # Not all features require both, but for simplicity we expect both
            pass

    def find_open_projects(self) -> List[Dict[str, Any]]:
        if not self.db_projects:
            return []
        query = {
            "database_id": self.db_projects,
            "filter": {
                "property": "Status",
                "select": {"equals": "Open"}
            }
        }
        res = self.client.databases.query(**query)
        items = []
        for page in res.get("results", []):
            props = page.get("properties", {})
            name = props.get("Name", {}).get("title", [])
            title = "".join([t.get("plain_text", "") for t in name]) if name else ""
            status = props.get("Status", {}).get("select", {}).get("name", "")
            division = [x.get("name") for x in props.get("Divisi", {}).get("multi_select", [])]
            desc = props.get("Deskripsi", {}).get("rich_text", [])
            description = "".join([t.get("plain_text", "") for t in desc]) if desc else None
            apply_url = props.get("Link Pendaftaran", {}).get("url")
            items.append({
                "id": page.get("id"),
                "name": title,
                "status": status,
                "division": division,
                "description": description,
                "apply_url": apply_url,
            })
        return items

    def _get_block_text_recursive(self, block_id: str) -> str:
        texts: List[str] = []
        cursor = None
        while True:
            res = self.client.blocks.children.list(block_id=block_id, start_cursor=cursor)
            for blk in res.get("results", []):
                t = blk.get("type")
                rich = blk.get(t, {}).get("rich_text", [])
                if rich:
                    texts.append("".join([r.get("plain_text", "") for r in rich]))
                # recurse if has children
                if blk.get("has_children"):
                    texts.append(self._get_block_text_recursive(blk.get("id")))
            if not res.get("has_more"):
                break
            cursor = res.get("next_cursor")
        return "\n".join([x for x in texts if x])

    def get_document(self, doc_name: str) -> Optional[str]:
        if not self.db_docs:
            return None
        query = {
            "database_id": self.db_docs,
            "filter": {
                "property": "Name",
                "title": {"equals": doc_name}
            }
        }
        res = self.client.databases.query(**query)
        results = res.get("results", [])
        if not results:
            return None
        page = results[0]
        page_id = page.get("id")
        content = self._get_block_text_recursive(page_id)
        return content

    def get_document_by_id(self, page_id: str) -> Optional[str]:
        if not page_id:
            return None
        return self._get_block_text_recursive(page_id)

    def list_documents(self, category: Optional[str] = None, page_size: int = 100, include_content: bool = True) -> List[Dict[str, Any]]:
        if not self.db_docs:
            return []
        payload: Dict[str, Any] = {
            "database_id": self.db_docs,
            "page_size": page_size,
        }
        if category:
            payload["filter"] = {
                "property": "Kategori",
                "select": {"equals": category}
            }
        docs: List[Dict[str, Any]] = []
        cursor = None
        while True:
            if cursor:
                payload["start_cursor"] = cursor
            res = self.client.databases.query(**payload)
            for page in res.get("results", []):
                props = page.get("properties", {})
                name = props.get("Name", {}).get("title", [])
                title = "".join([t.get("plain_text", "") for t in name]) if name else "Untitled"
                page_id = page.get("id")
                last_edited_time = page.get("last_edited_time")
                item: Dict[str, Any] = {
                    "id": page_id,
                    "title": title,
                    "last_edited_time": last_edited_time,
                }
                if include_content:
                    item["content"] = self._get_block_text_recursive(page_id)
                docs.append(item)
            if not res.get("has_more"):
                break
            cursor = res.get("next_cursor")
        return docs
